[7.0] SEPA SWISS LOCALIZATION

abaghirli opened the pull request.

Commit:
SEPA SWISS LOCALIZATION

yvaucher commented on l10n_ch_sepa_credit_transfer/__init__.py:
```diff
@@ -0,0 +1,22 @@
+##############################################################################
+#
+#    Author: Yannick Vaucher
+#    Copyright 2011 Camptocamp SA
```
You can change this to put you as author and 2017.

yvaucher commented on l10n_ch_sepa_credit_transfer/__openerp__.py:
```diff
@@ -0,0 +1,57 @@
+
+It currently supports the "pain.001" norm for Credit Transfert Initiation.
+
+""",
```
This description needs to be moved in README.rst

abaghirli replied:
in version 7 there is no Readme.rst files in modules.

yvaucher commented on l10n_ch_sepa_credit_transfer/__openerp__.py:
```diff
@@ -0,0 +1,57 @@
+##############################################################################
+#
+#    Author: Yannick Vaucher
+#    Copyright 2011 Camptocamp SA
```
You can change author and year here

yvaucher commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,242 @@
+                'sepa_export': sepa_export,
+                'file_obj': self.pool['banking.export.sepa'],
+                'pain_xsd_file':
+                'account_banking_sepa_credit_transfer/data/%s.xsd'
```
Here you can do:

``` python
if pain_flavor ...:
   module = ...
else:
   module = ...
gen_args = {
            [...]
            'pain_xsd_file': '%s/data/%s.xsd' % (module, pain_flavor)
}
```

yvaucher commented on l10n_ch_sepa_direct_debit/wizard/export_sdd.py:
```diff
@@ -0,0 +1,337 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    SEPA Credit Transfer module for OpenERP
```
s/Credit Transfer/Direct Debit

yvaucher commented on l10n_ch_sepa_direct_debit/wizard/export_sdd.py:
```diff
@@ -0,0 +1,337 @@
+    _inherit = 'banking.export.sdd.wizard'
+
+
+def create_sepa(self, cr, uid, ids, context=None):
```
Indentation is wrong here

yvaucher commented on l10n_ch_sepa_direct_debit/wizard/export_sdd.py:
```diff
@@ -0,0 +1,337 @@
+            'file_obj': self.pool['banking.export.sdd'],
+            'pain_xsd_file':
+                'account_banking_sepa_direct_debit/data/%s.xsd' % pain_flavor,
+        }
```
Same comment than for credit transfer. Here you could have a placeholder for module

yvaucher reviewed (changes requested):
I only see XSD validation changes. Are some commit missing ? It seems that an equivalence to 
https://github.com/OCA/l10n-switzerland/tree/10.0/l10n_ch_pain_base
is missing.


Indent is wrong in l10n_ch_sepa_direct_debit/wizard/export_sdd.py

Commit:
requested changes

intend and xml code

intend

yvaucher commented on l10n_ch_sepa_credit_transfer/data/payment_type_sepa_sct.xml:
```diff
@@ -0,0 +1,19 @@
+<data noupdate="0">
+
+
+<!-- TODO In the suitable_bank_types field, we should restrict to IBAN type and Rib+IBAN of l10n_fr_rib ? -->
```
Not sure to get that comment, l10n_fr_rib is related to france localization not switzerland one.

yvaucher commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,231 @@
+            control_sum_1_7.text = '%.2f' % amount_control_sum_1_7
+        else:
+            nb_of_transactions_1_6.text = str(transactions_count_1_6)
+            control_sum_1_7.text = '%.2f' % amount_control_sum_1_7
```
If and else are doing the exact same thing?

abaghirli replied:
i agree. But i was afraid to remove this part of code:
https://github.com/OCA/bank-payment/blob/7.0/account_banking_sepa_credit_transfer/wizard/export_sepa.py#L278

yvaucher commented on l10n_ch_sepa_direct_debit/__openerp__.py:
```diff
@@ -0,0 +1,57 @@
+
+{
+    "name": "Switzerland - SEPA Electronic Payment File",
+    "summary": "Generate pain.001 Credit Transfert Files for your payments",
```
Name and summary to fix.

yvaucher commented on l10n_ch_sepa_direct_debit/__openerp__.py:
```diff
@@ -0,0 +1,57 @@
+{
+    "name": "Switzerland - SEPA Electronic Payment File",
+    "summary": "Generate pain.001 Credit Transfert Files for your payments",
+    "version": "1.0",
```
version should be 7.0.1.0.0

yvaucher commented on l10n_ch_sepa_direct_debit/__openerp__.py:
```diff
@@ -0,0 +1,57 @@
+    "author": "Camptocamp,Odoo Community Association (OCA)",
+    "depends": [
+        "account",
+        "l10n_ch",
```
We might not need to depend on l10n_ch as it is only the chart of account.

yvaucher commented on l10n_ch_sepa_direct_debit/data/payment_type_sepa_sct.xml:
```diff
@@ -0,0 +1,19 @@
+<data noupdate="0">
+
+
+<!-- TODO In the suitable_bank_types field, we should restrict to IBAN type and Rib+IBAN of l10n_fr_rib ? -->
```
Same comment as in other module.
This is related to question about france localisation.

yvaucher reviewed (changes requested):
Please check travis for lint

Trying manually, I cannot generate a credit transfer:

Here is the error I get:

```
Payment Type Code 'pain.001.001.03.ch.02' is not supported. The only Payment Type Codes supported for SEPA Credit Transfers are 'pain.001.001.02', 'pain.001.001.03', 'pain.001.001.04', 'pain.001.001.05', 'pain.001.003.03' and 'pain.001.003.03.ch.02'. 
```

yvaucher commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,231 @@
+            bic_xml_tag = 'BIC'
+            name_maxsize = 70
+            root_xml_tag = 'CstmrCdtTrfInitn'
+        elif pain_flavor == 'pain.001.003.03.ch.02':
```
I guess there is a typo here:

xml data file is defining: `pain.001.001.03.ch.02`

yvaucher commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,231 @@
+        # A. Group header
+        group_header_1_0, nb_of_transactions_1_6, control_sum_1_7 = \
+            self.generate_group_header_block(
+                cr, uid, pain_root, gen_args, context=context)
```
Here, `gen_args` is undefined if pain_flavor == 'pain.001.003.03.ch.02'

yvaucher commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,231 @@
+            module = 'l10n_ch_sepa_credit_transfer'
+        else:
+            module = 'account_banking_sepa_credit_transfer'
+            gen_args = {
```
Indentation is also wrong here

Commit:
requested chenges, version number and etc

xml tag fix

unclosed tags and bank account number

lint

simahawk commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,303 @@
+        pain_flavor = sepa_export.payment_order_ids[0].mode.type.code
+        convert_to_ascii = \
+            sepa_export.payment_order_ids[0].mode.convert_to_ascii
+        if pain_flavor == 'pain.001.001.02':
```
these are all static values: move them to a mapping. For instance:
```python
class BankingExportSepaWizard(...):
    pain_flavor_mapping = {
        'pain.001.001.02':  {...}
        'pain.001.001.03':  {...}
        [...]
    }
    [...]
    pain_values = self.pain_flavor_mapping.get(pain_flavor)
```

simahawk commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,303 @@
+                _('Error:'),
+                _("Payment Type Code '%s' is not supported. The only "
+                    "Payment Type Codes supported for SEPA Credit Transfers "
+                    "are 'pain.001.001.02', 'pain.001.001.03', "
```
according to the change request above:
```python
"are %s" % ', '.join(self.pain_flavor_mapping.keys())
```

simahawk commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,303 @@
+        # key = (requested_date, priority)
+        # values = list of lines as object
+        today = fields.date.context_today(self, cr, uid, context=context)
+        for payment_order in sepa_export.payment_order_ids:
```
split this to a separate method `group_lines`

simahawk commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,303 @@
+                else:
+                    lines_per_group[key] = [line]
+                # Write requested_date on 'Payment date' of the pay line
+                if requested_date != line.date:
```
if we expect many lines here is better to collect a list of `(line_id, date)` to write everything at the end w/ a single query.
See [this example](https://github.com/camptocamp/bso_odoo/blob/master/odoo/local-src/sale_purchase_sourcing/wizards/wiz_sale_order_source.py#L285).

simahawk commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,303 @@
+
+            transactions_count_2_4 = 0
+            amount_control_sum_2_5 = 0.0
+            for line in lines:
```
split this to a method too and pass it only required args like the counters

simahawk commented on l10n_ch_sepa_credit_transfer/wizard/export_sepa.py:
```diff
@@ -0,0 +1,303 @@
+                eval_ctx, gen_args, context=context)
+        return True
+
+    def create_sepa(self, cr, uid, ids, context=None):
```
:stop_sign:  this method is way too long: more than 200 lines. Hard to read, understand and test.
See my suggestions for splitting it into helper methods.

simahawk commented on l10n_ch_sepa_direct_debit/wizard/export_sdd.py:
```diff
@@ -0,0 +1,398 @@
+                eval_ctx, gen_args, context=context)
+        return True
+
+    def create_sepa(self, cr, uid, ids, context=None):
```
same as above: way too long. Should be organized into meaningful blocks delegated to helper methods.

yvaucher replied:
@simahawk this is base on a port of a module from version 8 or 9 we probably don't want to change it too much.

simahawk commented:
BTW, no tests?

Commit:
priority tag

fixed tags and xsd

company issuer tag

lint

lint 2

payment 1 2.1 Ok

payment type 2.2 OK

flake8

Commit:
clearing check
